aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp

```diff
@@ -0,0 +1,206 @@
+      x_qparams = fbgemm::ChooseQuantizationParams(
+          x_min_data[i],
+          x_max_data[i],
+          qmin,
+          qmax,
+          symmetric_quant,
+          /* preserve sparsity */ // Set preserve_sparsity to true for
+                                  // symmetric quantization.
+          false /* force power of two*/);
+      scale[i] = x_qparams.scale;
+      zero_point[i] = x_qparams.zero_point;
+#else
+      quant_utils::TensorQuantizationParams x_qparams{};
+      x_qparams = quant_utils::ChooseQuantizationParams(
+          x_min_data[i],
+          x_max_data[i],
+          qmin,
+          qmax,
+          symmetric_quant, /* preserve sparsity */
+          false /* force power of two*/);
+      scale[i] = x_qparams.scale;
+      zero_point[i] = x_qparams.zero_point;
+#endif
+    }
+    fake_quant_out = at::fake_quantize_per_channel_affine_cachemask(
+        x, scale, zero_point, ch_axis, qmin, qmax);
+  } else {
+#ifdef USE_FBGEMM
+    fbgemm::TensorQuantizationParams x_qparams{};
+    // compute quantization parameters using min-max values
+    x_qparams = fbgemm::ChooseQuantizationParams(
+        inp_running_min.item().toFloat(),
+        inp_running_max.item().toFloat(),
+        qmin,
+        qmax,
+        symmetric_quant, /* bool preserve_sparsity */
+        false /* force power of two*/);
+
+    scale[0] = x_qparams.scale;
+    zero_point[0] = x_qparams.zero_point;
+#else
+    quant_utils::TensorQuantizationParams x_qparams{};
+    // compute quantization parameters using min-max values
+    x_qparams = quant_utils::ChooseQuantizationParams(
+        inp_running_min.item().toFloat(),
+        inp_running_max.item().toFloat(),
+        qmin,
+        qmax,
+        symmetric_quant, /* bool preserve_sparsity */
+        false /* force power of two*/);
+    scale[0] = x_qparams.scale;
+    zero_point[0] = x_qparams.zero_point;
+#endif
+    fake_quant_out =
+        at::_fake_quantize_per_tensor_affine_cachemask_tensor_qparams(
+            x, scale, zero_point, qmin, qmax);
+  }
+  return fake_quant_out;
+}
+} // namespace
+
+namespace at {
+namespace native {
+
+std::tuple<at::Tensor, at::Tensor> fused_moving_avg_obs_fake_quant_cpu(
+    const at::Tensor& self,
+    const at::Tensor& observer_on,
+    const at::Tensor& fake_quant_on,
+    const at::Tensor& averaging_const,
+    at::Tensor& running_min,
+    at::Tensor& running_max,
+    at::Tensor& scale,
+    at::Tensor& zero_point,
+    const int64_t quant_min,
+    const int64_t quant_max,
+    const int64_t ch_axis,
+    bool per_row_fake_quant,
+    bool symmetric_quant) {
+  // Calculate min/max
+  auto observe = observer_on.item().toInt();
+  if (observe) {
+    calculate_moving_average(
+        self,
+        averaging_const,
+        running_min,
+        running_max,
+        per_row_fake_quant,
+        ch_axis);
+  }
+  // Calculate qparams and fake_quantize
+  auto fake_quant = fake_quant_on.item().toInt();
+  if (fake_quant) {
```

Comment:
You probably want to factor your helper function so that it never returns a view by moving the `if fake_quant` logic in to the composite op instead (`fused_moving_avg_obs_fake_quant`).

The reason why you needed the don't enforce skip in `gen_variable_type` is because you return an alias here but did not declare the it as a view in `gen_inplace_or_view`. However, there are also situations e.g., when `fake_quant` is false, where you return a copy, so declaring this function as a view isn't entirely correct either.

So the rule in general is that a non-composite op should either consistently return a view or non-view. Factoring the original op into a composite wrapper + helper in this case allows us to shift that view/non-view inconsistency into the composite op, which skips the autograd kernel and thus forgoes the check for storage and tensor use count.

Reply:
Since this involved creating multiple helper functions and break down of composite vs non-composite ops, I'm returning `x.clone()` instead when `fake_quant` is off. Running some jobs now to see if there is any impact due to this.